
# A pulmonologist’s guide to perform and analyse cross-species single-lung-cell transcriptomics 

Single-cell ribonucleic acid sequencing (scRNA-seq) is becoming widely employed to study biological processes in great depth. Its name describes its key advantage, the resolution of transcriptomes at single-cell level.
In the Publication mentioned above,we exemplary demonstrate current procedures and tools to create, compare and qualitatively assess the single-cell transcriptomes of the lung in six different species; human (Homo sapiens), African green monkeys (Chlorocebus sabaeus), pigs (Sus domesticus), hamster (Mesocricetus auratus), rats (Rattus norvegicus), and mice (Mus musculus). We used quality-filtered scRNA-seq data of lungs from different sources using up to date analysis workflows.
All cross-species analysis code of this example including expression of canonical celltype-specific genes is made available via this Github page. For the creation shown Figures further processing and alignment was performed with Inkscape (Inkscape Community; https://inkscape.org/de/; gitlab.com/inkscape/inkscape). 

Input files as well as the processed, integrated result file are available as Seurat-R-object at the [Leipzig Health Atlas](https://www.health-atlas.de/studies/54). 

![grafik](https://user-images.githubusercontent.com/73164857/157898202-fda5e47f-9f26-41f5-a6c3-8a59e3e44956.png)

FIGURE 2
Cross-species comparison of sequenced lung cells. A) Uniform manifold approximation and projection (UMAP) plot of identified cell populations across species with pie charts of lung cell frequencies in indicated species estimated from scRNA-Seq data. B) Dot plots indicating the stress response of lung cells across species based on the 8 stress related genes HSPA8, FOS, DUSP1, IER3, EGR1, FOSB, HSPB1 and ATF3. AM: alveolar macrophages, Prol. AM: proliferating alveolar macrophages, Mɸ: interstitial macrophages/monocytes, PMN: polymorphonuclear leukocytes, DC: dendritic cells, SMC: Smooth muscle cells, Mast: mast cells, NK: natural killer cells, Prol. NK/T: proliferating NK and T cells, AT1: alveolar epithelial cells type 1, AT2: alveolar epithelial cells type 2, Ciliated: ciliated cells, Club: club cells, EC: endothelial cells, ly EC: lymphatic endothelial cells, Fibro: fibroblasts, Perivasc: perivascular cells. Samples: Human Charité, cells from four fresh lung explants. Human Travaglini et al., data of patient 2 (https://doi.org/10.1038/s41586-020-2922-4). Mouse, cells of two whole lungs pooled prior analysis. Hamster, cells from lobus caudalis of three hamsters, data from https://doi.org/10.1038/s41467-021-25030-7. Monkey cells from two lungs, data from https://doi.org/10.1126/scitranslmed.abe8146. Rat and Pig cells from two lungs each, data from https://doi.org/10.1126/sciadv.aaw3851. All experiments involving animals were approved by institutional and governmental authorities (Charité Universitätsmedizin Berlin or Freie Universität Berlin and LaGeSo Landesamt für Gesundheit und Soziales Berlin, Germany). Patient Material of Human Charité: The study was approved by the ethics committee at the Charité clinic (projects EA2/079/13), written informed consent was obtained from all patients. 


# Supplementary Figures to the manuscript 
![grafik](https://user-images.githubusercontent.com/73164857/157898462-06d44001-2409-4eee-8cec-c7ee1c51571a.png)

Supplementary Figure 2. Overview of top gene markers used for cell identification and their expression strength between species. Dotplot of cross-species comparison for top RNA marker genes. The point size indicates the percentage of cells in a certain population expressing the indicated marker gene. Average expression is indicated by colour intensity while species is indicated by colouration. 

# Supplementary Informations on sample material

Human Charité: Dataset consists of lung cells isolated from fresh lung explants of four 65 - 85 years old patients (3 male, 1 female), suffering from lung carcinoma who underwent lung resection. Only tumor-free peripheral lung tissue was used. The study was approved by the ethics committee at the Charité Universitätsmedizin Berlin (projects EA2/079/13). Written informed consent was obtained from all patients. Single cell barcoding and library construction was done according to the manufacturer’s instructions (10X Genomics, run in four experimental batches). Alignment and quantification was done using cellranger against the human GRCh38 genome. Standard Seurat workflow was performed, excluding cells with 10% or more mitochondrial genes and less than 500 genes, batch effect with were removed using Seurat::IntegrateData(), and doublets were identified using [scrublet](https://github.com/swolock/scrublet). The dataset includes expression information for 44621 expression features for a total of 18487 cells.

Human Travaglini et al.: Sequenced lung cell data of patient 2, processed via 10X Genomics and Smart-Seq2 workflow, published at https://www.synapse.org by Travaglini et al. (https://doi.org/10.1038/s41586-020-2922-4) was used. According to the original publication the cells were isolated from freshly resected lung tissue of a 46-year-old male. The dataset used here includes expression information for 26485 expression features for a total of 28793 cells originating from eight experimental batches.

Monkey: Datset of two monkey lungs (mock infected controls), published at https://www.ncbi.nlm.nih.gov/geo by Speranza et al. (https://doi.org/10.1126/scitranslmed.abe8146) was used.

Hamster: Dataset consists of lung cells isolated from lobus caudalis of two female and one male Syrian hamsters (Mesocricetus auratus; breed RjHan:AURA, Janvier Labs, France) at 10 to 12 weeks of age. It includes expression information for 35604 expression features for a total of 10564 cells resulting from three experimental batches. All experiments involving animals were approved by institutional and governmental authorities (Freie Universität Berlin and LaGeSo Landesamt für Gesundheit und Soziales Berlin, Germany). For further Information see: https://doi.org/10.1038/s41467-021-25030-7

Mouse: Dataset consists of pulmonary cells isolated from whole lung tissue. The single cell suspension of two 8 to 10 weeks old C57BL/6J mice was pooled prior performing single cell barcoding and library construction according to manufacturer’s instructions (10X Genomics). All experiments involving animals were approved by institutional and governmental authorities (Charité Universitätsmedizin Berlin and LaGeSo Landesamt für Gesundheit und Soziales Berlin, Germany). Alignment and quantification was done using cellranger against the human GRCh38 genome. Low-quality cells were removed by filtering out cells not having less than 1000 and not more than 35000 RNA counts quantified as unique molecular identifier (UMIs), not having at least 300 different genes expressed and not having more than 30% mitochondrial genes expressed. Data was normalized applying standard Seurat 4.0.1 SCTransform workflow, the mouse data was available as a single batch. The dataset includes expression information for 45046 expression features for a total of 3205 cells.

Rat and Pig: Datsets of two lungs each, published at https://www.ncbi.nlm.nih.gov/geo by Raredon et al. (https://doi.org/10.1126/sciadv.aaw3851) were used.
